pkg/messaging/mqtt/subscriber.go

```diff
@@ -16,21 +16,58 @@ import (
 )
 
 var (
-	errSubscribeTimeout   = errors.New("failed to subscribe due to timeout reached")
-	errUnsubscribeTimeout = errors.New("failed to unsubscribe due to timeout reached")
-	errAlreadySubscribed  = errors.New("already subscribed to topic")
-	errNotSubscribed      = errors.New("not subscribed")
-	errEmptyTopic         = errors.New("empty topic")
-	errEmptyID            = errors.New("empty ID")
+	errSubscribeTimeout       = errors.New("failed to subscribe due to timeout reached")
+	errUnsubscribeTimeout     = errors.New("failed to unsubscribe due to timeout reached")
+	errUnsubscribeDeleteTopic = errors.New("failed to unsubscribe due to deletion of topic")
+	errAlreadySubscribed      = errors.New("already subscribed to topic")
+	errNotSubscribed          = errors.New("not subscribed")
+	errEmptyTopic             = errors.New("empty topic")
+	errEmptyID                = errors.New("empty ID")
 )
 
 var _ messaging.Subscriber = (*subscriber)(nil)
 
+type subscription struct {
+	client mqtt.Client
+	topics []string
+}
+
+// contains checks if a topic is present
+func (sub subscription) contains(topic string) bool {
+	for _, v := range sub.topics {
+		if v == topic {
+			return true
+		}
+	}
+	return false
+}
+
+// Finds the index of an item in the topics
+func (sub subscription) indexOf(element string) int {
+	for k, v := range sub.topics {
+		if element == v {
+			return k
+		}
+	}
+	return -1
+}
+
+// Deletes a topic from the slice
+func (sub subscription) delete(topic string) bool {
+	index := sub.indexOf(topic)
+	if index == -1 {
+		return false
+	}
+	copy(sub.topics[index:], sub.topics[index+1:])
+	sub.topics[len(sub.topics)-1] = ""
```

Comment:
Probably the simplest way to handle slice growth is to actually copy it. Something like:
```go
topics := make([]string, len(sub.topics)-1)
copy(topics[:index], sub.topics[:index])
copy(topics[index:], sub.topics[index+1:])
sub.topics = topics
```

Reply:
Please double-check this implementation - since you're only copying, it looks like it does not actually free any space. You have an example here: https://github.com/mainflux/mainflux/pull/1598/files#r863116725.
It is especially dangerous because this check will, eventually, never be satisfied: https://github.com/mainflux/mainflux/pull/1598/files#diff-3740ddbdbe88f161342d4cb97fb90d933e7ac87450cb30dcb246002912230e95R153.